### PR TITLE
Create documents in merlin thread

### PR DESCRIPTION
**Pull request by rgrinberg**
It does not seem to be safe to do these operations concurrently to
merlin

### CODE REVIEW

**Comment by rgrinberg** [comment]
@ulugbekna will wait for explicit approval this time. Sorry for the impatience.

**Review by ulugbekna — APPROVED** [review]
Looks good to me. 

Rebased and made some renamings. Feel free to remove my commit if you don't agree with it -- thought it's better to make changes myself than bothering you.

**Pull request merged**
